README.md
# Base45

[![Test and publish Base45](https://github.com/opendevtools/base45/workflows/Release/badge.svg?branch=main)](https://github.com/opendevtools/base45/actions?query=workflow%3A%22Release%22)
[![Crate](https://img.shields.io/crates/v/base45.svg)](https://crates.io/crates/base45)
[![API](https://docs.rs/base45/badge.svg)](https://docs.rs/base45)

A encoder/decoder for base45 that is fully compatible with
[draft-faltstrom-base45-02](https://www.ietf.org/id/draft-faltstrom-base45-02.txt). When encoding QR or Aztec codes a different scheme than the standard base64, base32, and base16 is needed.

## Installation

```toml
[dependencies]
base45 = "1.0.1"
```

## Benchmarks

Benchmarks were created using [Criterion](https://github.com/bheisler/criterion.rs). The benchmarks test encoding and decoding of the string "The quick brown fox jumps over the lazy dog". Test computer is a MacBook Pro 14" (2021) M1 Pro with 32 GB RAM.

```
encode long string              time:   [107.51 ns 107.69 ns 107.91 ns]
encode long string from buffer  time:   [126.66 ns 126.84 ns 127.07 ns]
decode long string              time:   [115.58 ns 115.81 ns 116.08 ns]

encode long str 100 times       time:   [12.805 µs 12.821 µs 12.839 µs]
decode long str 100 times       time:   [12.407 µs 12.424 µs 12.445 µs]

encode 0x10 random bytes        time:   [57.717 ns 57.827 ns 57.946 ns]
encode 0x100 random bytes       time:   [292.00 ns 292.64 ns 293.50 ns]
encode 0x1000 random bytes      time:   [3.6926 µs 3.6962 µs 3.7001 µs]
encode 0x10000 random bytes     time:   [65.210 µs 65.476 µs 65.755 µs]
```


CHANGELOG.md
# [3.1.0](https://github.com/opendevtools/base45/compare/v3.0.0...v3.1.0) (2023-06-30)


### Features

* `array_chunks` (performance feature), `alloc` improvements ([#3](https://github.com/opendevtools/base45/issues/3)) ([9e0e933](https://github.com/opendevtools/base45/commit/9e0e9333243f87c485c32ef2ed208ed3a1d509f2))
* performance improvements ([#2](https://github.com/opendevtools/base45/issues/2)) ([63db6f2](https://github.com/opendevtools/base45/commit/63db6f2495a593d7df7acb506716687d8fef3033))

# [3.0.0](https://github.com/opendevtools/base45/compare/v2.0.1...v3.0.0) (2021-06-03)

### Features

- return decoding errors ([#1](https://github.com/opendevtools/base45/issues/1)) ([185d311](https://github.com/opendevtools/base45/commit/185d311951e0f567dc9ccfd4c2cbf9219271ce45))

### BREAKING CHANGES

- Update the return value of `base45::decode` to `Result<_,_>`. This allows us to handle errors gracefully instead of panicking

## [2.0.1](https://github.com/opendevtools/base45/compare/v2.0.0...v2.0.1) (2021-04-02)

### Bug Fixes

- handle invalid strings ([c9957b5](https://github.com/opendevtools/base45/commit/c9957b5f43e8df798bb31086ea29c27452333e79))

# [2.0.0](https://github.com/opendevtools/base45/compare/v1.0.2...v2.0.0) (2021-04-01)

### Features

- add encode_from_buffer ([f8b3892](https://github.com/opendevtools/base45/commit/f8b3892987a13f9fb2376ccb78adfb00d1e71950))
- decode returns buffer ([dca8ddf](https://github.com/opendevtools/base45/commit/dca8ddf956df2219aea07e926928409a5426863f))

### BREAKING CHANGES

- decode used to return a string, now it returns a
  buffer and the user should convert it to a string

## [1.0.2](https://github.com/opendevtools/base45/compare/v1.0.1...v1.0.2) (2021-03-22)

### Bug Fixes

- create release zip ([2587c27](https://github.com/opendevtools/base45/commit/2587c272f21b6eb7f02685a1841cfc280aeba53a))

## [1.0.1](https://github.com/opendevtools/base45/compare/v1.0.0...v1.0.1) (2021-03-22)

### Bug Fixes

- update docs ([28b8e00](https://github.com/opendevtools/base45/commit/28b8e006d315d73d445b969ffedb37b8705d6164))

# 1.0.0 (2021-03-22)

### Bug Fixes

- remove public size ([b263d2f](https://github.com/opendevtools/base45/commit/b263d2fdb28015b3c16d162b1a010a8619a08a17))

### Features

- initial commit ([c6389d7](https://github.com/opendevtools/base45/commit/c6389d774e77b814684d965e3f26433e28b7f4c6))


src/tests.rs
use crate::*;

#[test]
fn encode_ab() {
    assert_eq!(encode("AB"), "BB8")
}

#[test]
fn decode_fail() {
    assert!(decode(":::").is_err());
}

#[test]
fn decode_fail_out_of_range() {
    assert!(decode(":::").is_err());
}

#[test]
fn encode_hello() {
    assert_eq!(encode("Hello!!"), "%69 VD92EX0")
}

#[test]
fn encode_base45() {
    assert_eq!(encode("base-45"), "UJCLQE7W581")
}

#[test]
fn encode_long_string() {
    assert_eq!(
        encode("The quick brown fox jumps over the lazy dog"),
        "8UADZCKFEOEDJOD2KC54EM-DX.CH8FSKDQ$D.OE44E5$CS44+8DK44OEC3EFGVCD2",
    )
}

#[test]
fn encode_unicode() {
    assert_eq!(encode("foo © bar 𝌆 baz"), "X.C82EIROA44GECH74C-J1/GUJCW2")
}

#[test]
fn encode_emoji() {
    assert_eq!(encode("I ❤️  Rust"), "0B98TSD%K.ENY244JA QE")
}

#[test]
fn encode_ietf() {
    assert_eq!(encode("ietf!"), "QED8WEX0")
}

#[test]
fn decode_ab() {
    assert_eq!(String::from_utf8(decode("BB8").unwrap()).unwrap(), "AB")
}

#[test]
fn decode_hello() {
    assert_eq!(
        String::from_utf8(decode("%69 VD92EX0").unwrap()).unwrap(),
        "Hello!!"
    )
}

#[test]
fn decode_base45() {
    assert_eq!(
        String::from_utf8(decode("UJCLQE7W581").unwrap()).unwrap(),
        "base-45"
    )
}

#[test]
fn decode_ietf() {
    assert_eq!(
        String::from_utf8(decode("QED8WEX0").unwrap()).unwrap(),
        "ietf!"
    )
}

const QUICK_BROWN_FOX_ENC: &str =
    "8UADZCKFEOEDJOD2KC54EM-DX.CH8FSKDQ$D.OE44E5$CS44+8DK44OEC3EFGVCD2";
const QUICK_BROWN_FOX_DEC: &str = "The quick brown fox jumps over the lazy dog";
#[test]
fn decode_long_string() {
    assert_eq!(
        String::from_utf8(decode(QUICK_BROWN_FOX_ENC).unwrap()).unwrap(),
        QUICK_BROWN_FOX_DEC,
    )
}

#[test]
fn encode_hello_from_buffer() {
    assert_eq!(encode(vec![72, 101, 108, 108, 111, 33, 33]), "%69 VD92EX0")
}

#[test]
fn encode_full_bytes() {
    let s = encode(b"\xff\xff\xff\xff\xff\xff");
    assert_eq!(s, "FGWFGWFGW");
    let s = encode(b"\xff\xff\xff\xff\xff\xff\xff");
    assert_eq!(s, "FGWFGWFGWU5");
    let s = encode(b"\xff\xff\xff\xff\xff\xff\xff\xff");
    assert_eq!(s, "FGWFGWFGWFGW");
}
#[test]
fn decode_full_bytes() {
    let s = decode("FGWFGWFGW").unwrap();
    assert_eq!(s, b"\xff\xff\xff\xff\xff\xff");
    let s = decode("FGWFGWFGWU5").unwrap();
    assert_eq!(s, b"\xff\xff\xff\xff\xff\xff\xff");
    let s = decode("FGWFGWFGWFGW").unwrap();
    assert_eq!(s, b"\xff\xff\xff\xff\xff\xff\xff\xff");
}

// cursed code, quickly becomes degenerate.

// fn rbd<const N: usize>(bench: &mut test::Bencher) {
//     use rand::{distributions::*, *};
//     use std::convert::TryFrom;
//     let mut rng = thread_rng();
//     let sample = Slice::new(&crate::alphabet::TABLE).unwrap();
//     let b: Vec<u8> = sample.sample_iter(rng).copied().take(N).collect();
//     let b: [u8; N] = TryFrom::try_from(b).unwrap();
//     bench.iter(|| {
//         let decoded = decode(&b[..]);
//         assert!(decoded.is_ok());
//     });
// }
// #[bench]
// fn bench_decode_random_3(b: &mut test::Bencher) {
//     rbd::<3>(b);
// }
// #[bench]
// fn bench_decode_random_30(b: &mut test::Bencher) {
//     rbd::<30>(b);
// }
// #[bench]
// fn bench_decode_random_3000(b: &mut test::Bencher) {
//     rbd::<3000>(b);
// }
// #[bench]
// fn bench_decode_random_3002(b: &mut test::Bencher) {
//     rbd::<3002>(b);
// }


src/lib.rs
#![cfg_attr(feature = "array_chunks", feature(array_chunks))]
#![cfg_attr(test, feature(test))]
//! Encoder/decoder for base45 that is fully compatible with
//! [`draft-faltstrom-base45-07.txt`](https://www.ietf.org/archive/id/draft-faltstrom-base45-07.txt)
//!
//! ```rust,no_run
//! # fn main() {
//!     let encoded = base45::encode("Hello!!");
//! # }
//! ```
//!
//! Features:
//!
//! - `array_chunks`, which is using experimental array and const-generic features. For information & tracking, see
//!   [rust/rust#74985](https://github.com/rust-lang/rust/issues/74985). If not enabled, this uses
//!   [`core::slice::ChunksExact`](https://doc.rust-lang.org/core/slice/struct.ChunksExact.html).
//!   Ideally, there is no performance penalty using this means.

pub mod alphabet;
mod decode;
mod encode;

pub use decode::*;
pub use encode::*;

#[cfg(test)]
mod tests;


src/decode.rs
use crate::alphabet::{self, SIZE, SIZE_SIZE};
use std::error::Error;
use std::fmt::{Display, Formatter};
/// The error type returned when the input is not a valid base45 string
#[derive(Eq, PartialEq, Copy, Clone, Debug)]
pub struct DecodeError;

impl Display for DecodeError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_str("Invalid base45 string")
    }
}

impl Error for DecodeError {}

/// Decode a string from base45
///
/// Takes a base45 encoded string and returns a UTF-8 string on success
///
/// ```rust,no_run
/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
/// let decoded = String::from_utf8(base45::decode("%69 VD92EX0")?)?;
/// assert_eq!(decoded, "Hello!!");
/// # Ok(())
/// # }
/// ```
pub fn decode(input: impl AsRef<[u8]>) -> Result<Vec<u8>, DecodeError> {
    decode_intl(input.as_ref())
}
/// Internal function to decode a string from base45, to reduce code bloat.
fn decode_intl(input: &[u8]) -> Result<Vec<u8>, DecodeError> {
    let mut output = Vec::with_capacity(match input.len() & 1 {
        0 => input.len() * 2 / 3,
        1 => 1 + input.len() * 2 / 3,
        // SAFETY: it's one of 0 or 1. There are no other options.
        _ => unsafe { core::hint::unreachable_unchecked() },
    });

    #[inline(always)]
    fn core_fn([_first, _second, _third]: [u8; 3]) -> Result<[u8; 2], DecodeError> {
        let v = (_first as u32) + (_second as u32 * SIZE) + (_third as u32) * SIZE_SIZE;
        if v <= (u16::MAX as _) {
            let x = (v >> 8) & 0xFF;
            let y = v & 0xFF;
            Ok([x as u8, y as u8])
        } else {
            Err(DecodeError)
        }
    }
    //short
    #[inline(always)]
    fn preproc([a, b, c]: [u8; 3]) -> Result<[u8; 3], DecodeError> {
        Ok([d(a)?, d(b)?, d(c)?])
    }
    //short
    #[inline(always)]
    fn d(v: u8) -> Result<u8, DecodeError> {
        alphabet::decode(v).ok_or(DecodeError)
    }

    #[cfg(feature = "array_chunks")]
    let (chunks, pre) = (input.array_chunks(), |&b| preproc(b));
    #[cfg(not(feature = "array_chunks"))]
    let (chunks, pre) = (input.chunks_exact(3), |slic: &[u8]| match slic {
        &[a, b, c] => preproc([a, b, c]),
        // SAFETY: chunks_exact ensures every `.next()` returns an effective &[T] where `.len` == 3
        _ => unsafe { core::hint::unreachable_unchecked() },
    });

    let remainder = chunks.remainder();
    for c in chunks.map(pre) {
        let xy = core_fn(c?)?;
        output.extend_from_slice(&xy);
    }
    match remainder {
        [_first, _second] => output.push((d(*_first)? as u32 + (d(*_second)? as u32 * SIZE)) as u8),
        [] => {}
        _ => return Err(DecodeError),
    }

    Ok(output)
}


src/alphabet.rs
/// Table of characters (by index). Currently unused.
pub const TABLE: [u8; 45] = *b"0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ $%*+-./:";

/// Decode a byte to its index.
pub const fn decode(v: u8) -> Option<u8> {
    // table statements are fast, but use more space. Enabled by default.
    Some(match v {
        b'0' => 0,
        b'1' => 1,
        b'2' => 2,
        b'3' => 3,
        b'4' => 4,
        b'5' => 5,
        b'6' => 6,
        b'7' => 7,
        b'8' => 8,
        b'9' => 9,
        b'A' => 10,
        b'B' => 11,
        b'C' => 12,
        b'D' => 13,
        b'E' => 14,
        b'F' => 15,
        b'G' => 16,
        b'H' => 17,
        b'I' => 18,
        b'J' => 19,
        b'K' => 20,
        b'L' => 21,
        b'M' => 22,
        b'N' => 23,
        b'O' => 24,
        b'P' => 25,
        b'Q' => 26,
        b'R' => 27,
        b'S' => 28,
        b'T' => 29,
        b'U' => 30,
        b'V' => 31,
        b'W' => 32,
        b'X' => 33,
        b'Y' => 34,
        b'Z' => 35,
        b' ' => 36,
        b'$' => 37,
        b'%' => 38,
        b'*' => 39,
        b'+' => 40,
        b'-' => 41,
        b'.' => 42,
        b'/' => 43,
        b':' => 44,
        _ => return None,
    })
}

pub fn encode(n: u8) -> Option<u8> {
    //#[cfg(feature = "bin_table")]
    TABLE.get(n as usize).map(|&c| c)
    // very fast, literal index pointer
}

pub const SIZE: u32 = 45;
pub const SIZE_SIZE: u32 = SIZE * SIZE;


benches/bench.rs
use criterion::{black_box, criterion_group, criterion_main, Criterion};

fn encode(c: &mut Criterion) {
    c.bench_function("encode long string", |b| {
        b.iter(|| base45::encode(black_box("The quick brown fox jumps over the lazy dog")))
    });
}

fn encode_from_buffer(c: &mut Criterion) {
    c.bench_function("encode long string from buffer", |b| {
        b.iter(|| {
            base45::encode_from_buffer(black_box(vec![
                84, 104, 101, 32, 113, 117, 105, 99, 107, 32, 98, 114, 111, 119, 110, 32, 102, 111,
                120, 32, 106, 117, 109, 112, 115, 32, 111, 118, 101, 114, 32, 116, 104, 101, 32,
                108, 97, 122, 121, 32, 100, 111, 103,
            ]))
        })
    });
}

fn decode(c: &mut Criterion) {
    c.bench_function("decode long string", |b| {
        b.iter(|| {
            String::from_utf8(
                base45::decode(black_box(
                    "8UADZCKFEOEDJOD2KC54EM-DX.CH8FSKDQ$D.OE44E5$CS44+8DK44OEC3EFGVCD2",
                ))
                .unwrap(),
            )
            .unwrap()
        })
    });
}

criterion_group!(benches, encode, encode_from_buffer, decode);
criterion_main!(benches);


src/encode.rs
use crate::alphabet::{self, SIZE, SIZE_SIZE};
fn divmod<const N: u32>(x: u32) -> (u32, u32) {
    (x / N, x % N)
}

fn ae(b: u8) -> u8 {
    match alphabet::encode(b) {
        Some(ch) => ch,
        // SAFETY: encode for this is highly unlikely to ever reach this point.
        #[cfg(not(test))]
        None => unsafe { core::hint::unreachable_unchecked() },
        #[cfg(test)]
        None => unreachable!(),
    }
}

fn encode_buffer(input: &[u8]) -> String {
    // setup
    #[cfg(feature = "array_chunks")]
    let input = input.array_chunks::<2>();
    #[cfg(not(feature = "array_chunks"))]
    let input = input.chunks_exact(2);

    let mut s = Vec::with_capacity(input.len() + ((input.len() + 1) / 2));

    // Core function
    #[inline(always)]
    fn core_fn([_first, _second]: [u8; 2], s: &mut Vec<u8>) {
        let v = (_first as u32 * 256) + _second as u32;
        let (e, rest) = divmod::<SIZE_SIZE>(v);
        let (d, c) = divmod::<SIZE>(rest);

        s.extend_from_slice(&[ae(c as _), ae(d as _), ae(e as _)]);
    }
    // take remainder AoT
    let rem = input.remainder();
    for c in input {
        #[cfg(feature = "array_chunks")]
        let c = *c;
        // SAFETY: `chunks_exact` always returns exactly that number of items.
        #[cfg(not(feature = "array_chunks"))]
        let c = unsafe { [*c.get_unchecked(0), *c.get_unchecked(1)] };
        core_fn(c, &mut s);
    }
    // Final byte
    if let &[_0] = rem {
        let (d, c) = divmod::<SIZE>(_0 as u32);

        s.extend_from_slice(&[ae(c as _), ae(d as _)]);
    }
    // SAFETY: we control all bytes that enter this vector.
    unsafe { String::from_utf8_unchecked(s) }
}

/// Encode a string to base45
///
/// The function takes a string containing only characters in
/// in the range U+0000 to U+00FF.
///
/// ```rust,no_run
/// # fn main() {
/// let encoded = base45::encode("Hello!!");
/// assert_eq!(encoded, "%69 VD92EX0");
/// # }
/// ```
pub fn encode(input: impl AsRef<[u8]>) -> String {
    encode_buffer(input.as_ref())
}

/// Encode a buffer to base45
///
/// The function takes an arbitrary buffer and encodes it to base45.
///
/// ## Updates
///
/// As of v3.1.0, this function is exactly eqivalent to encode.
///
/// ```rust,no_run
/// # fn main() {
/// let encoded = base45::encode_from_buffer(vec![72,101,108,108,111,33,33]);
/// assert_eq!(encoded, "%69 VD92EX0");
/// # }
/// ```
#[deprecated(
    since = "3.1.0",
    note = "Equivalent to `encode`. Use `encode` instead."
)]
pub fn encode_from_buffer(input: impl AsRef<[u8]>) -> String {
    encode_buffer(input.as_ref())
}
